### PR TITLE
Rule expansion

### PR DESCRIPTION
**Pull request by einari**
This PR lays the groundwork for new capabilities for working with rules. 
And also hardens it with more specs.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1115616103797671/1150102418739672)

### CODE REVIEW

**Comment by einari** [comment]
STOP THE PRESS..   
@smithmx had a great idea, and we'll do it - we'll rename BrokenRuleReason *ProximateCause* and BrokenRuleReasonInstance to *ActualCause*.

File: Source/Rules/BrokenRule.cs [hunk 1]
```diff
@@ -29,28 +29,28 @@ public BrokenRule(IRule rule, object instance, IRuleContext context)
...
         /// Gets the context in which the rule was broken
         /// </summary>
-        public IRuleContext Context { get; private set; }
+        public IRuleContext Context { get; }
 
         /// <summary>
         /// Gets the <see cref="BrokenRuleReason">reasons</see> why the rule is broken
         /// </summary>
-        public IEnumerable<BrokenRuleReason> Reasons { get { return _reasons; } }
+        public IEnumerable<BrokenRuleReasonInstance> Reasons => _reasons;
```
Comment (smithmx) [inline Source/Rules/BrokenRule.cs]: Should we do _reasons.ToArray() ?  Just so the external state isn't exposed.  I don't worry about it too much as if you use reflection to discover it's a list and add to it... well, you deserve all you get.  It's not too much further to get access to a private variable via reflection.

What was my point again?

File: Source/Rules/RuleContext.cs [hunk 2]
```diff
@@ -13,16 +13,28 @@ public class RuleContext : IRuleContext
...
         public void OnFailed(RuleFailed callback)
         {
             _callbacks.Add(callback);
         }
 
-        public void Fail(IRule rule, object instance, BrokenRuleReason reason)
+        /// <inheritdoc/>
+        public void Fail(IRule rule, object instance, BrokenRuleReasonInstance reason)
         {
             _callbacks.ForEach(c => c(rule, instance, reason));
```
Comment (smithmx) [inline Source/Rules/RuleContext.cs]: That's just cruel... 

I'm here all week.  Remember to tip your waitresses.

File: Source/Rules/RuleEvaluationResult.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *--------------------------------------------------------------------------------------------*/
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Dolittle.Rules
+{
+    /// <summary>
+    /// Represents the result of a an evaluation of a <see cref="IRule"/>
+    /// </summary>
+    public class RuleEvaluationResult
```
Comment (smithmx) [inline Source/Rules/RuleEvaluationResult.cs]: I've got a feeling we might want to generalize this.  I'd much prefer to do this than return a null.

File: Source/Rules/RuleSetEvaluation.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        /// <summary>
+        /// Implicitly convert to <see cref="bool"/> for evaluating the success of the evaluation
+        /// </summary>
+        /// <param name="input"><see cref="RuleSetEvaluation"/> to convert from</param>
+        public static implicit operator bool(RuleSetEvaluation input) => input.IsSuccess;
+
+        /// <summary>
+        /// Gets whether or not the <see cref="RuleSetEvaluation"/> was successful or not
+        /// </summary>
+        public bool IsSuccess => _brokenRules.Count == 0;
```
Comment (smithmx) [inline Source/Rules/RuleSetEvaluation.cs]: !_brokenRules.Any();

File: Source/Rules/BrokenRule.cs (hunk 1 above)
Comment (einari) [reply Source/Rules/BrokenRule.cs]: Actually from an immutability perspective, that is a good pattern to stick to.

File: Source/Rules/RuleContext.cs (hunk 2 above)
Comment (einari) [reply Source/Rules/RuleContext.cs]: 🤣

**Pull request merged**